### PR TITLE
fix(timestamp): add trim for the input date string

### PR DESCRIPTION
**Pull request by etolbakov**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
`trim` the input date string  before parsing the timestamp

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link
https://github.com/GreptimeTeam/greptimedb/issues/2069

### CODE REVIEW

File: src/query/src/optimizer/string_normalization.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+struct StringNormalizationConverter;
+
+impl TreeNodeRewriter for StringNormalizationConverter {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        let new_expr = match expr {
+            Expr::Cast(Cast { expr, data_type }) => {
```
Comment (etolbakov) [inline src/query/src/optimizer/string_normalization.rs:52]: maybe we need an extra check here that `data_type` is Timestamp

File: tests/cases/standalone/common/timestamp/timestamp.sql [hunk 2]
```diff
@@ -1,3 +1,5 @@
+SELECT timestamp '    2017-07-23    13:10:11   ';
+
```
Comment (etolbakov) [inline tests/cases/standalone/common/timestamp/timestamp.sql:2]: I've updated [dummy.sql](https://github.com/GreptimeTeam/greptimedb/blob/d7002caca75f7123d155dbb0eb621e37b46b887a/tests/cases/standalone/common/select/dummy.sql#L1) / [dummy.result](https://github.com/GreptimeTeam/greptimedb/blob/d7002caca75f7123d155dbb0eb621e37b46b887a/tests/cases/standalone/common/select/dummy.result#L1) and have the following result 

```
[dummy.sql]
select TO_UNIXTIME('    2023-03-01T06:35:02Z    ');
...
[dummy.result]
select TO_UNIXTIME('    2023-03-01T06:35:02Z    ');

+---------------------------------------------------+
| to_unixtime(Utf8("    2023-03-01T06:35:02Z    ")) |
+---------------------------------------------------+
| 1677652502                                        |
+---------------------------------------------------+
``` 
Not sure if that's a valid usecase though. Let me know if you'd like this to be added

Comment (waynexia) [reply tests/cases/standalone/common/timestamp/timestamp.sql:2]: This looks fine to me 👍

File: tests/cases/standalone/common/timestamp/timestamp.sql [hunk 3]
```diff
@@ -1,3 +1,5 @@
+SELECT timestamp '    2017-07-23    13:10:11   ';
```
Comment (waynexia) [inline tests/cases/standalone/common/timestamp/timestamp.sql:1]: How about also adding `SELECT ' 2017-07-23 13:10:11 '` to make sure the normal string literal is not affected

File: src/query/src/query_engine/state.rs [hunk 4]
```diff
@@ -84,6 +85,7 @@ impl QueryEngineState {
             analyzer.rules.insert(0, Arc::new(DistPlannerAnalyzer));
         }
         analyzer.rules.insert(0, Arc::new(TypeConversionRule));
+        analyzer.rules.insert(1, Arc::new(StringNormalizationRule));
```
Comment (waynexia) [inline src/query/src/query_engine/state.rs:88]:
```suggestion
        analyzer.rules.insert(0, Arc::new(StringNormalizationRule));
```

They have no order dependency so just insert into the first position

File: src/query/src/optimizer/string_normalization.rs (hunk 1 above)
Comment (waynexia) [reply src/query/src/optimizer/string_normalization.rs:52]: Yes please

File: src/query/src/optimizer/string_normalization.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    use datafusion_expr::{lit, Cast, Expr, LogicalPlanBuilder};
+    use datafusion_optimizer::analyzer::AnalyzerRule;
+
+    use crate::optimizer::string_normalization::StringNormalizationRule;
+
+    #[test]
+    fn test_string_normalization_with_extra_spaces() {
+        let proj = vec![Expr::Cast(Cast::new(
+            Box::new(lit("    2017-07-23    13:10:11   ")),
+            DataType::Timestamp(Nanosecond, None),
```
Comment (waynexia) [inline src/query/src/optimizer/string_normalization.rs:96]: Expect another case that casts to data types other timestamp types

File: src/query/src/optimizer/string_normalization.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            .analyze(plan, &ConfigOptions::default())
+            .unwrap();
+        let expected = String::from(
+               "Projection: CAST(Utf8(\"2017-07-23 13:10:11\") AS Timestamp(Nanosecond, None))\n  TableScan: t"
+        );
+        assert_eq!(expected, format!("{:?}", result));
+    }
+
+    fn prepare_test_plan_builder() -> LogicalPlanBuilder {
+        let batch = RecordBatch::try_from_iter_with_nullable(vec![
```
Comment (etolbakov) [inline src/query/src/optimizer/string_normalization.rs:114]: I will try simplifying this.

**Comment by etolbakov** [comment]
@waynexia 
Could you please take a look when you have time?

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**